kitty/model/low_level/encoder.py

```diff
@@ -84,53 +107,38 @@ def __init__(self, func):
         self._func = func
 
     def encode(self, value):
-        kassert.is_of_types(value, str)
-        encoded = self._func(value)
+        kassert.is_of_types(value, (bytes, str))
+        encoded = self._func(strToBytes(value))
         return Bits(bytes=encoded)
 
 
 class StrEncodeEncoder(StrEncoder):
```

Comment:
change the class to subclass StrFuncEncoder. I'm posting my change here because I'm not sure how to edit yours:
```suggestion
@@ -112,7 +112,7 @@ class StrFuncEncoder(StrEncoder):
         return Bits(bytes=encoded)


-class StrEncodeEncoder(StrEncoder):
+class StrEncodeEncoder(StrFuncEncoder):
     '''
     Encode the string using str.encode function
     '''
@@ -122,24 +122,17 @@ class StrEncodeEncoder(StrEncoder):
         :type encoding: ``str``
         :param encoding: encoding to be used, should be either one of 'hex', 'base64', 'utf-8', 'bytes' or  valid argument StrFuncEncoder
         '''
-        super(StrEncodeEncoder, self).__init__()
         if encoding == 'hex':
-            self._func = hexlify
+            func = hexlify
         elif encoding == 'base64':
-            self._func = b64encode
+            func = b64encode
         elif encoding == 'utf-8':
-            self._func = strToUtf8
+            func = strToUtf8
         elif encoding == 'bytes':
-            self._func = strToBytes
-        self._encoding = encoding
-
-    def encode(self, value):
-        '''
-        :param value: value to encode
-        '''
-        kassert.is_of_types(value, (bytes, str))
-        encoded = StrFuncEncoder(self._func).encode(value)
-        return Bits(bytes=encoded)
+            func = strToBytes
+        else:
+            func = encoding
+        super(StrEncodeEncoder, self).__init__(func)
```

Reply:
But I didn't sure, removing `encode()` function from Encoder may be best solution?

Reply:
The encode function should be removed from `StrEncodeEncoder`, but I think the code I pasted was not displayed as I meant it. what I meant was this:

```python
class StrEncodeEncoder(StrFuncEncoder):
    '''
    Encode the string using str.encode function
    '''

    def __init__(self, encoding):
        '''
        :type encoding: ``str``
        :param encoding: encoding to be used, should be a valid argument for str.encode
        '''
        if encoding == 'hex':
            func = hexlify
        elif encoding == 'base64':
            func = b64encode
        elif encoding == 'utf-8':
            func = strToUtf8
        elif encoding == 'bytes':
            func = strToBytes
        else:
            func = encoding
        super(StrEncodeEncoder, self).__init__(func)
```

This way you don't instantiate a new encoder for each call to encode() and you get a much simpler class.

Reply:
Yes you are right.